### PR TITLE
Iron fold contracts

### PR DESCRIPTION
**Pull request by CryptJS13**
Altered all contracts to be compatible with 0.8.4 and made them fit your interfacing and general structure. Everything compiles, but have not gotten to testing yet.

### CODE REVIEW

File: contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "../StrategyBase.sol";
+import "../../../third_party/iron/IIronChef.sol";
+import "../../../third_party/iron/IRMatic.sol";
+import "../../../third_party/iron/CompleteRToken.sol";
+
+
+/// @title Abstract contract for Iron strategy implementation
+/// @author belbix
```
Comment (belbix) [inline contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol:25]: Jasper? :)

File: contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+abstract contract IronFoldStrategyBase is StrategyBase {
```
Comment (belbix) [inline contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol:26]: we will have two strategies, one for LP pairs and this for lending
I guess need to reflect it in the name

File: contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+  }
+
+  /// @notice Contract constructor using on strategy implementation
+  /// @dev The implementation should check each parameter
+  /// @param _controller Controller address
+  /// @param _underlying Underlying token address
+  /// @param _vault SmartVault address that will provide liquidity
+  /// @param __rewardTokens Reward tokens that the strategy will farm
+  /// @param _rToken RToken address
+  /// @param _ironController Iron Controller address
```
Comment (belbix) [inline contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol:65]: add descriptions for the new attributes

File: contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+    address _vault,
+    address[] memory __rewardTokens,
+    address _rToken,
+    address _ironController,
+    uint256 _borrowTargetFactorNumerator,
+    uint256 _collateralFactorNumerator,
+    uint256 _factorDenominator,
+    bool _fold
+  ) StrategyBase(_controller, _underlying, _vault, __rewardTokens, _BUY_BACK_RATIO) {
+    require(_rToken != address(0), "zero address pool");
```
Comment (belbix) [inline contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol:78]: zero rToken I guess

File: contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+  string public constant VERSION = "1.0.0";
+  /// @dev Placeholder, for non full buyback need to implement liquidation
+  uint256 private constant _BUY_BACK_RATIO = 10000;
+
+  /// @notice RToken address
+  address public rToken;
+  /// @notice Iron Controller address
+  address public ironController;
+
+  uint256 public borrowTargetFactorNumerator;
```
Comment (belbix) [inline contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol:44]: please describe in comments these variables purposes

File: contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+    uint256[] memory rewards = new uint256[](1);
+    return rewards;
+  }
+
+  // ************ GOVERNANCE ACTIONS **************************
+
+  /// @notice Claim rewards from external project and send them to FeeRewardForwarder
+  function doHardWork() external onlyNotPausedInvesting override restricted {
+    claimReward();
+    liquidateReward();
```
Comment (belbix) [inline contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol:130]: maybe call rebalance for each hard work?

File: contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol (hunk 1 above)
Comment (CryptJS13) [reply contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol:25]: Right haha

File: contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol (hunk 2 above)
Comment (CryptJS13) [reply contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol:26]: I thought Fold would be enough, no folding on LPs right?

File: contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol (hunk 6 above)
Comment (CryptJS13) [reply contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol:130]: Good idea

File: contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol (hunk 3 above)
Comment (CryptJS13) [reply contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol:70]: Done

File: contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol (hunk 5 above)
Comment (CryptJS13) [reply contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol:45]: Done

File: contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol (hunk 2 above)
Comment (belbix) [reply contracts/base/strategies/iron-fold/IronFoldStrategyBase.sol:26]: ah sure, my bad

**Pull request merged**